### ISSUE TITLE
Use CastleRandom in the Engine (formerly: TGenericStructList shuffle?)

### ISSUE THREAD

**Issue by eugeneloza** [castle-engine/castle-engine#64]
Do we need to have a shuffle procedure in a TGenericStructList and similar? I'm speaking of http://forum.lazarus.freepascal.org/index.php/topic,34882.msg229682.html#msg229682 . Never needed anything like that myself yet. However, looks like it's a nice feature worth adding and looks like a simple thing to do (maybe, even make helpers for normal generic lists). If "yes" then I'm on it :)

**Comment by michaliskambi** [comment]
Hm, I would not add it for now. If one of us will need it in an actual game, then I'm all after adding it. But not earlier:)

Reasoning: There are probably hundreds of such "nice to have" methods for a generic list, and implementing them all (without an actual use case)... well, it seems like we could use our time better, to implement something immediately useful:)

That said, it will not make maintenance harder or anything, it's a simple method, so I'm not really against it. So, I leave the decision up to you.

**Comment by michaliskambi** [comment]
I'll leave it open, up to you:)

**Comment by eugeneloza** [comment]
I nearly missed one just a week ago, still solved the problem with a different method. Plus, I've been awfully lazy last week (due to external factors, like illness :)), so I want to do something small but useful :) And I've seen several people asking for and using this approach.
P.S. We don't have any kind of random enabled globally for the Engine? So, basically there are three ways we can do it:
1. Using System.Random. Not optimal, e.g. I don't randomize it due to using CastleRandom.
2. Using an internal unit-wise random. Doesn't interfere with other randoms. But slows down initialization and consumes RAM.
3. Accept a `@random` function as a parameter. Looks like the best variant, because one can use his own initialized TCastleRandom generator to get repeatable random sequences. However, it's not "universal" as we'll accept only internal Castle Random, with, probably, falling back system.random (i.e. default/overload) if `nil` is provided.

**Comment by michaliskambi** [comment]
The existing engine randomness routines (e.g. in CastleUtils) use System.Random right now. It is automatically randomized at the initialization of CastleUtils, which is at the bottom of (almost?) every other CGE unit.

So, this is the one you should use (at least by default).

I think that AD 2 and AD 3 are a bit over-complications of a simple task... unless, like in my previous comment, it's something you actually have a use for. No other randomization routines allow configuring their randomness source for now.

As for AD 2: if you want to choose this variant, I would bring my older suggestion (I think I voiced it earlier): define a `Rnd` function in `CastleRandom` unit that returns a global TCastleRandom  instance created on-demand (so doesn't slow down the initialization, yet is comfortable to use).

Similar to the `SoundEngine` function in `CastleSoundEngine` unit:

~~~~
var
  FSoundEngine: TRepoSoundEngine;

function SoundEngine: TRepoSoundEngine;
begin
  if FSoundEngine = nil then
    FSoundEngine := TRepoSoundEngine.Create;
  Result := FSoundEngine;
end;

finalization
  FreeAndNil(FSoundEngine);
end.
~~~~

But, I emphasize, I would choose AD 1 and not complicate the design more:)

**Comment by eugeneloza** [comment]
Let's add another question to the heap :)
I've found a use for "random based on a string" for one of my projects (not the main one, and one probably would be never developed actively) - Oazis game (a huge openworld game with proceduraly-generated plot (by fractal-like algorithms) set in Decoherence universe / I still haven't decided whether it'll be 3D/2D or a text adventure (if it'll ever be at all)).
E.g. we have a NPC stat "Metal crafting skill" (among thousands of others, like "Afraid of mice") so the NPC has a basic seed and XOR seed determined by Random(stat_name) = some random value which will always be the same for the same NPC and the same stat (so there is no need to save stats values for a million of NPCs unless they are very close friends and therefore their character may be influenced by the player character). More specific, this seed will "guide" a special procedure to determine the value (as some values are linked like "Profession: trader" and "Trade skill" or "Race: Chameleon" and "Skill: Stealth"). Even more specific: each situation/dialogue will "request(specific_stat)" in a specified list, and "request" will check 1) if the stat is saved else 2) if stat is "remarkable" then generate remarkable stat else 3) generate a normal stat.
The algorithm is simple. XOR all bytes of the string (in groups of 4 to get 32 bit accuracy) and xorshift it once = seed based on the string. Then this seed might XOR with another seed or used as a random initialization.
Yes, "String_to_Seed" doesn't seem like and often-used function :) However, a catchphrase "My new world" as a seed might do a better job, than "5491709754".

**Comment by michaliskambi** [comment]
IOW, you calculate a (simple) hash value of a string, and then use it as an integer random seed. Sounds cool to me, and your example ("My new world" instead of "5491709754") is nice:)

If you're asking: "Can we add it to CastleRandom", then I say "Yes, if you want to do it":) "StringToRandomSeed" seems like a good name to me :)

I checked, and I don't think there's a simple string-hashing function (string->32 bit value) in FPC available already. Although there are various more advanced algorithms in packages/hash/src (MD5, SHA1 etc.), but they don't produce a simple 32-bit value. The http://www.freepascal.org/docs-html/fcl/contnrs/tfphashlist.html class has an internal string->int hashing function, but it's not publicly available. For your interest, I paste it below (it's in FPC sources in packages/fcl-base/src/contnrs.pp):

~~~~
    Function FPHash(const s:shortstring):LongWord;
    var
      p,pmax : PChar;
    begin
{$push}
{$Q-}
      Result:=0;
      p:=@s[1];
      pmax:=@s[length(s)+1];
      while (p<pmax) do
        begin
          Result:=LongWord(LongInt(Result shl 5) - LongInt(Result)) xor LongWord(P^);
          Inc(p);
        end;
{$pop}
    end;

    Function FPHash(P: PChar; Len: Integer): LongWord;
    var
      pmax : PChar;
    begin
{$push}
{$Q-}
      Result:=0;
      pmax:=p+len;
      while (p<pmax) do
        begin
          Result:=LongWord(LongInt(Result shl 5) - LongInt(Result)) xor LongWord(P^);
          Inc(p);
        end;
{$pop}
    end;
~~~~

**Comment by michaliskambi** [comment]
I'm wondering, maybe we should close this issue?

To be clear: I don't mind keeping this issue open, and it's OK for an issue to be open for a long time... but only if we *really* plan to solve it eventually :) As I don't see an immediate need for this feature, I don't plan to implement it myself. So the question is left up to you --- do you know that you will need this feature in one of your projects?

If not, maybe we can close this? We can always reopen this if the need will arise :)

**Comment by eugeneloza** [comment]
Hi, Michalis!
Sorry for the delay. I hope I'll get to normal lifeflow soon :) I plan to add this feature asap, but still recovering from Ph.D. defense. But I hope eventually not only to return to normal work flow, but also to take a vacation to finish all the planned stuff like this :)

I still need it open to keep the list of tasks I plan to do. But... hmm... I don't see any way to move the issues to my fork (that'd be a good place to keep 'em, and to keep the master branch clean). The only way I've found is https://github-issue-mover.appspot.com/, but I don't think it's a good idea to trust them.

**Comment by michaliskambi** [comment]
It's OK then -- if you have plans to solve this, then I have no objections to having this issue open :)

**Comment by eugeneloza** [comment]
Ahh... why didn't I think about that before? Still too hard to think straight. Now I see it's easily done.
I just make a reference to the closed issue in my fork and I get both a to-do list and we can close the issues (all that are pending), preserving easy access to all the messages I need to implement the specific stuff.

Moving the issue to my fork. Closed this one for now. I'll report here as soon as I've finished.

**Comment by michaliskambi** [comment]
OK :) I'll watch your repository too, so I'll comment if I have anything useful to add.

P.S. My personal TODO lists for CGE are maintained as a collection of text files (in my private SVN repo). There are 61 files with various TODO things now, some with many TODOs inside (the largest one is probably "non_urgent/improvements.org.txt" with 195 TODO items now). Maybe this will help, if you ever think that you have "too many unfinished TODOs" :)

**Comment by eugeneloza** [comment]
Spent quiet some time today on hashing...

FPHASH:
![2017-07-27-232248_1366x768_scrot](https://user-images.githubusercontent.com/10726162/28690881-7330fc26-7323-11e7-8cbf-c955d50b554f.png)
(Looks horrible)

My hashing with bit shuffling
![2017-07-27-232300_1366x768_scrot](https://user-images.githubusercontent.com/10726162/28690888-7bb118d6-7323-11e7-9f42-df1fe7a56d20.png)
(Looks bad)

My hashing with bit shuffling and xorshift
![2017-07-27-232312_1366x768_scrot](https://user-images.githubusercontent.com/10726162/28690909-92e1ebf2-7323-11e7-996f-3a08e830380b.png)
(also looks bad)

hmmm... I don't want to use any overly complex hashing algorithm, but something understandable to acuquire an even distribution which will produce significant difference if string changes by a single character.
Still working on it.

**Comment by eugeneloza** [comment]
MurMur2 looks fantastic...
![2017-07-28-000923_1366x768_scrot](https://user-images.githubusercontent.com/10726162/28692439-34e04e3a-7329-11e7-93e7-4717c00f4297.png)
However, I'm not sure about it, because it causes numeric overflows / range checking errors. This can cause unstable program behaviour, right?

**Comment by michaliskambi** [comment]
(I'm sorry for the very long answer! I wanted to explain it fully, and got carried away. I hope the explanation is useful in any case. The answer to your question is at the bottom!)

> This can cause unstable program behaviour, right?

Overflow and range check errors are, to put it simply, bugs.

- Overflow (checked by {$Q+}) is when you make an arithmetic operation that results in something larger than the allowed value. For example, I have `A: Int64` that is equal right now `High(Int64) - 100`. I calculate `A + 200`. The result is `High(Int64) + 100`, it cannot fit within the `Int64` range.
  
    I made a demo showing it: https://gist.github.com/michaliskambi/ea4399e9aeb9f23858d22dc58b1391d9

- Range check error (checked by {$R+}) is when you have e.g. `Arr: array[0..10] of something` and you access (read or write) `Arr[11]` or `Arr[-1]`.

- To confuse matters a bit, if you try to repeat my 1st example with Byte (8 bit unsigned integer), declaring `A: Byte` and initializing `A := 100` at the beginning and you calculate `A + 200`... (which is 300, and Byte can only fit values 0..255)... you will get a *range check error*, not *overflow error*  as you might expect. The reason for this inconsistency is somewhat internal.

    (FPC calculates internally using native integer size, 32 or 64 bit on our modern machines, and only an error at this point (at 32-bit or 64-bit calculation) causes *overflow*; an error later, because the result doesn't fit in Byte range, causes a *range check error*. See Jonas comments at https://bugs.freepascal.org/view.php?id=28384 )

    Fortunately, in most usecases you enable (or disable) both overflow and range checking, so it doesn't matter much which check ($Q or $R) catches which error (overflow at arithmetic operation, or accessing outside of array bounds).

These errors are detected in debug mode, and cause an exception then. But this detection takes time (quite significant in some cases, e.g. our ray-tracer is 1.9 x slower in debug mode, see https://castle-engine.sourceforge.io/manual_optimization.php#section_release_mode ). So these checks are turned off in "release" mode. In release mode we simply assume you will not make these errors.

Now what will happen if these errors occur in release mode?

- An (unchecked) overflow means that the integer value will "wrap" around zero. It's easy to calculate what this means for unsigned values, when A is Byte and A = 200 and you do A := A + 100 then A is now equal (300 - 256) = 44. For signed values, you need to understand how negative integers are encoded in memory: https://en.wikipedia.org/wiki/Two%27s_complement . In any case, the result is probably something unexpected (you added something positive to A, but the resulting A is smaller) and can cause any logical bugs later.

- An (unchecked) range error, when you access something outside the array bounds, is much more nasty. You are reading outside of the memory than you're supposed to. 
    1. Maybe *you're reading a memory currently allocated but unused*, holding garbage (random garbage, that may even change from program execution to execution). 
    2. Maybe *you're reading a memory currently allocated and used* by another of your variables. So you read or write an unrelated, "innocent" variable. This can lead to many nasty bugs later. And it can be used as a security hole in special cases: an attacker can deliberately cause this error to read some adjacent memory location (e.g. containing user's saved passing).
    3. Maybe *you're reading outside of the allocated memory*. ("Allocated" from OS, using e.g. `malloc` from libc on Unix.) In this case you get SEGFAULT (a Unix name for Pascal's EAccessViolation exception).

So, to answer your question in short: *you have to write your code such that it never causes an overflow or range check error*. These mistakes cause clear exceptions in "debug" mode to help you with this (so that these errors don't go unnoticed). So you have to find in code what and why causes an overflow or range check error, and properly secure from it.

**Comment by eugeneloza** [comment]
Thanks for the explanations! Yes, I've heard that and it was exactly the thing that worried me (that overflow may cause instable behaviour).
But overflow is in the heart of MurMur2 algorithm... hmm... I should think a bit more about it. Looks like I can make everything 64 bit and clamp to 32 bit...

**Comment by eugeneloza** [comment]
Yes, writing "and $FFFFFFFF" did the job. Now a bit rewriting and tidying the code and it should work fine.

**Comment by michaliskambi** [comment]
(I created a wiki attached to the "modern-pascal-introduction" GitHub project, and I pasted there my ramblings about "what are overflow and range check errors". See https://github.com/michaliskambi/modern-pascal-introduction/wiki and https://github.com/michaliskambi/modern-pascal-introduction/wiki/What-are-range-and-overflow-checks-%28and-errors%29-in-Pascal ). Maybe it will be useful to someone in the future :)

**Comment by eugeneloza** [comment]
Yeah... I should update Russian translation also, it already lags behind in several chapters :) Plus there are several corrections I've made to English text which you might find useful.

**Comment by eugeneloza** [comment]
Done. It seems to work flawlessly :)
I wonder, should I place it in CastleRandom unit or maybe it might have some broader usage?
![2017-07-28-022045_1366x768_scrot](https://user-images.githubusercontent.com/10726162/28696022-9a717a78-733b-11e7-88dc-01c0c027fa0e.png)

```
{This is an implementation of MurMur2 hash algorithm
 to make a cryptographic-unsafe but highly uniform
 32bit hash from a given string very quickly.
 It may be used for better initialization of user defined random seeds
 (e.g. "My New World" sounds much better than 6592202398)
 or for better sorting/searching of strings
 s is the input string
 Seed is optional parameter enabling initialization
 of the algorithm seed different from default
 Result is unsigned 32-bit integer
 Warning: the hash is deterministic, but the result may be different
 depending on OS bit count or endianness (test required)}
function HashMurMur2(const InputString: AnsiString; const Seed: LongWord=0): LongWord;
var h, k: LongWord; //MurMur variables
    p: pointer;
    i: integer;
  procedure CycleHash(var x: LongWord); {$IFDEF SUPPORTS_INLINE}inline;{$ENDIF}
  const m = $5bd1e995; //MurMur "magic" cycling constant
        MaxLongWord = $FFFFFFFF;
  begin
    x := QWord(x*m) and MaxLongWord //prevent overflows during multiplication;
  end;
begin
  i := length(InputString);
  h := seed xor i; //init the deterministic seed
  p := @InputString[1]; //untyped pointer to the first character

  //cycle through all bytes of the string in 32 bit blocks
  while (i >= 4) do begin
    k := PLongWord(p)^;   //get next 4 bytes of data and process them
    CycleHash(k);
    k := k xor (k shr 24);
    CycleHash(k);

    CycleHash(h);
    h := h xor k;         //merge data into hash

    inc(p,4); //advance to next character
    dec(i,4); //to gain some speed we don't use p>pmax-4 check
  end;

  //upmix 0..3 final bytes of data to hash
  if i  = 3 then h := h xor (PByte(p+2)^ shl 16);
  if i >= 2 then h := h xor (PByte(p+1)^ shl 8);
  if i >= 1 then begin
                   h := h xor PByte(p)^;
                   CycleHash(h);
                 end;

  //and add a few final mixes
  h := h xor (h shr 13);
  CycleHash(h);
  h := h xor (h shr 15);

  Result := h;
end;
```

**Comment by eugeneloza** [comment]
> define a Rnd function in CastleRandom unit that returns a global TCastleRandom instance created on-demand (so doesn't slow down the initialization, yet is comfortable to use).

I was thinking about such approach a bit, but actually addressing RND might be a relatively often used task (e.g. in case of sorting/shuffling) - many times more often than accessing SoundEngine. So won't be checking each time `if rnd=nil` be a slowdown for time-critical tasks? Maybe initializing it in initialization section still would have a better efficiency trade-off? Or an additional "init random" aka "randomize" function which actually preforms all the tasks of creating and initialization of random generator be nice?
Another way to solve this issue is to make an extremely simple procedural (not class) type random generator based on the same algorithms which won't require creation of a class at all. However, it'll still require proper initialization (thou I can make it lower quality (like FPC does) and therefore much faster and therefore suitable for initialization - it won't add any lag at all).

**Comment by michaliskambi** [comment]
Cool! I think placing it in CastleRandom and making the procedure public is cool, this allows usage by any user code.

Minor note is to adjust the style of a comment a bit to match CGE conventions and PasDoc, like this:

```
{ Implementation of MurMur2 hash algorithm
  to make a cryptographic-unsafe but highly uniform
  32bit hash from a given string very quickly.
  It may be used for better initialization of user defined random seeds
  (e.g. "My New World" sounds much better than 6592202398)
  or for better sorting/searching of strings.
  Warning: the hash is deterministic, but the result may be different
  depending on CPU architecture or endianness (test required).

  @param(InputString The input string)
  @param(Seed Optional parameter enabling initialization
    of the algorithm seed different from default)
  @returns(Unsigned 32-bit integer)
}
```

- I added 1 more space between { and "Implementation", 
- and I used the @param and @returns PasDoc tags https://github.com/pasdoc/pasdoc/wiki/ParamReturnsRaisesTag , in the style used in CGE (the text like "The input string" stands on it's own and starts with capital letter).
- and I changed "OS bit count" to "CPU architecture".

Functionally:
- It would be nice to actually make it independent of CPU architecture -- looking at the code, it should actually be independent already?
- It would be nice then to add a simple test to tests/ , like `AssertEquals(6592202398, MurMur2('My New World'))`
- It's uneasy to find a big-endian machine nowadays for testing, but if you analyze the code and decide "it should work for big-endian too", and place the above test... then it will be automatically checked when updating the Debian package. They run our testsuite on various miscellaneous architectures, including big-endian :)
- The name should contain the word Hash, I think. Either `HashMurMur2` or `MurMur2Hash`.

**Comment by michaliskambi** [comment]
> So won't be checking each time if rnd=nil be a slowdown for time-critical tasks?

The time of such check should be absolutely irrelevant in all measurements. It's comparing a pointer-size number with zero, this comparison is definitely lighting fast (and is done by many things under the hood in our generated code anyway).

The other solutions all have some drawbacks:

- If you initialize the random in the initialization section, then
    1. It will take time, even if you don't use this Rnd, which prevents you from making a *really* good initial seed, and you worked a lot to make the initial seed good,
    2. And there's a risk that another's unit (from the engine) initialization section will be fired first, in case of unit inter-dependencies. So this initialization could not use this Rnd, it would be still `nil` then. This is admittedly an unlikely edge-case, but still it's better to avoid it if it's easy, like now :)

- Making an alternative lower-quality initialization means lower-quality.

- Making an alternative non-class implementation means duplicating a (non-trivial) logic.

I would definitely not worry about the speed of this comparison with `nil`, I don't believe it could have *any* effect on the speed of final code execution :) You can mark the `Rnd` function `inline` (calling function is likely more costly than comparison with pointer-sized zero), although I also *very* doubt that it will have *any* measurable effect on any real code.

In short, that would be an over-optimization IMHO :) Better keep the code simple. All we need is a 3 line `Rnd` function :)

**Comment by michaliskambi** [comment]
P.S. If you're looking how to add a test -- see at `castle-engine/tests/testcastletriangles.pas` source code, it's probably the simplest test unit right now. Just a class, with 1 method, with a straightforward checking "does this function, when executed on these hardcoded inputs, return correct (also hardcoded) outputs?". You use this unit in `test_castle_game_engine.lpr` and you're done :) Compile and run the tests by `./compile_console.sh` (in console) or just by opening and running test_castle_game_engine.lpi from Lazarus (GUI test runner).

**Comment by eugeneloza** [comment]
Eventually I've understood how to write tests :) Looks like it's tested and working as expected: https://github.com/castle-engine/castle-engine/pull/71/commits/5696475193435d6a52056b68ff222a0f1a11102d Also checked "invalid" versions, everything works properly and the errors are caught if any.
